src/v/archival/archiver_operations_impl.h

```diff
@@ -0,0 +1,211 @@
+    /// it's not performing any retries. The upload workflow should retry in
+    /// case of error. Otherwise the retries are invisible to the scheduler and
+    /// the workflow. The method can be used to upload different types of
+    /// objects. Not only segments. The type of object is encoded in the 'type'
+    /// parameter and the key is generic.
+    virtual ss::future<upload_result> upload_stream(
+      const cloud_storage_clients::bucket_name& bucket,
+      cloud_storage_clients::object_key key,
+      uint64_t content_length,
+      ss::input_stream<char> stream,
+      cloud_storage::upload_type type,
+      retry_chain_node& parent)
+      = 0;
+
+    /// Upload manifest to the cloud storage location
+    ///
+    /// The location is defined by the manifest itself
+    virtual ss::future<upload_result> upload_manifest(
+      const cloud_storage_clients::bucket_name& bucket,
+      const cloud_storage::base_manifest& manifest,
+      cloud_storage::remote_manifest_path path,
+      retry_chain_node& parent)
+      = 0;
+};
+
+struct prepared_segment_upload {
+    inclusive_offset_range offsets;
+    size_t size_bytes{0};
+    bool is_compacted{false};
+    cloud_storage::segment_meta meta;
+    ss::input_stream<char> payload;
+};
+
+/// Wrapper for the archival::segment_upload
+struct segment_upload_builder_api {
+    segment_upload_builder_api() = default;
+    segment_upload_builder_api(const segment_upload_builder_api&) = default;
+    segment_upload_builder_api(segment_upload_builder_api&&) = default;
+    segment_upload_builder_api& operator=(const segment_upload_builder_api&)
+      = default;
+    segment_upload_builder_api& operator=(segment_upload_builder_api&&)
+      = default;
+    virtual ~segment_upload_builder_api() = default;
+
+    virtual ss::future<result<std::unique_ptr<prepared_segment_upload>>>
+    prepare_segment_upload(
+      ss::shared_ptr<cluster_partition_api> part,
+      size_limited_offset_range range,
+      size_t read_buffer_size,
+      ss::scheduling_group sg,
+      model::timeout_clock::time_point deadline)
+      = 0;
+};
+
+inline std::ostream& operator<<(std::ostream& o, const cluster_partition_api&) {
+    return o << "cluster_partition_api";
+}
+inline std::ostream&
+operator<<(std::ostream& o, const cloud_storage_remote_api&) {
+    return o << "cloud_storage_remote_api";
+}
+inline std::ostream&
+operator<<(std::ostream& o, const cluster_partition_manager_api&) {
+    return o << "cluster_partition_manager_api";
+}
+inline std::ostream&
+operator<<(std::ostream& o, const segment_upload_builder_api&) {
+    return o << "segment_upload_builder_api";
+}
+
+/// Create archiver_operations_api instance with mock objects
+ss::shared_ptr<archiver_operations_api> make_archiver_operations_api(
+  ss::shared_ptr<cloud_storage_remote_api>,
+  ss::shared_ptr<cluster_partition_manager_api>,
+  ss::shared_ptr<segment_upload_builder_api>,
+  cloud_storage_clients::bucket_name);
```

Comment:
:+1: this looks great for testing

Reply:
Assuming that `make_archiver_operations_api` will be ultimately called from `application.cc`, how will the input shared ptrs be setup there, eg this argument `ss::shared_ptr<cloud_storage_remote_api>`? 

It looks like the shared ptrs are supposed to own `remote` etc, which are themselves services. Should these arguments be references instead of shared ptrs so they can work with sharded services which will be passed into this method?

Reply:
So to understand correctly, `cloud_storage_remote_api` will wrap the actual `remote` sharded service reference? Then it sounds fine. FWIW I had to do something very similar on my PR https://github.com/redpanda-data/redpanda/blob/2b5371ae2edba65dc6efd6e0ac97eebb02d3f86f/src/v/redpanda/application.cc#L1841-L1862 

Is `std::unique_ptr` not used because it won't work with the sharded service instantiation? I faced this issue on my PR when I initially tried unique_ptr for polymorphism, at some point unique_ptr copy is attempted during service start which doesn't compile.

Reply:
I think it makes sense to change this to unique_ptr but it's much easier to do in a followup PR. Otherwise it'll create too many merge conflicts.

Reply:
I think it might not work with unique_ptr, at least it did not work in my PR.